### PR TITLE
Cmm ns

### PR DESCRIPTION
**Pull request by chmorgan**
@kayoub5 this is a very early pull request that isn't actually ready to go in but it does start to build out the timestamp support.

### CODE REVIEW

File: SharpPcap/ICaptureDevice.cs [hunk 1]
```diff
@@ -166,6 +166,14 @@ public interface ICaptureDevice : IDisposable
         /// Return the pcap link layer value of an adapter. 
         /// </summary>
         PacketDotNet.LinkLayers LinkType { get; }
+
+        #region Timestamp
+        /// <summary>
+        /// Note: It isn't possible to set the precision on a device that has already
+        /// been activated. Please Open() the device with the desired precision.
+        /// </summary>
+        TimestampResolution TimestampPrecision { get; }
```
Comment (kayoub5) [inline SharpPcap/ICaptureDevice.cs:175]: `Precision` or `Resolution` ?

File: SharpPcap/LibPcap/LibPcapLiveDevice.cs [hunk 2]
```diff
@@ -164,6 +164,24 @@ public override void Open(DeviceConfiguration configuration)
                 LibPcapSafeNativeMethods.pcap_set_promisc(PcapHandle, (int)(mode & DeviceModes.Promiscuous));
                 LibPcapSafeNativeMethods.pcap_set_timeout(PcapHandle, configuration.ReadTimeout);
 
+                if(configuration.TimestampResolution != null)
+                {
+                    int retval = LibPcapSafeNativeMethods.pcap_set_tstamp_precision(PcapHandle, (int)configuration.TimestampResolution);
+                    if(retval != 0)
+                    {
+                        throw new InvalidOperationException("pcap_set_tstamp_precision() failed - return value " + retval);
```
Comment (kayoub5) [inline SharpPcap/LibPcap/LibPcapLiveDevice.cs:172]: You may want to change this once #220 is merged with new configuration failure handling

File: SharpPcap/LibPcap/PcapDevice.cs [hunk 3]
```diff
@@ -586,6 +586,51 @@ public virtual void SendPacket(byte[] p, int size)
         /// <param name="size">The number of bytes to send</param>
         public abstract void SendPacket(ReadOnlySpan<byte> p);
 
+        #region Timestamp
+        public virtual TimestampResolution TimestampPrecision
+        {
+            get => (TimestampResolution)LibPcapSafeNativeMethods.pcap_get_tstamp_precision(PcapHandle);
```
Comment (kayoub5) [inline SharpPcap/LibPcap/PcapDevice.cs:592]: what happens if device not open?

File: SharpPcap/TimestampType.cs [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+namespace SharpPcap
+{
+    /// <summary>
+    /// Source of the timestamp
+    /// See https://github.com/the-tcpdump-group/libpcap/blob/master/pcap/pcap.h#L498 for details
+    /// </summary>
+    public enum TimestampType : int
+    {
+        Host,
+        HostLowPrecision,
```
Comment (kayoub5) [inline SharpPcap/TimestampType.cs:11]: Assign the numbers
```suggestion
        HostLowPrecision = 1,
```

File: SharpPcap/WinDivert/WinDivertDevice.cs [hunk 5]
```diff
@@ -333,6 +333,13 @@ public virtual void StopCapture()
             }
         }
 
+        #region Timestamp
+        public virtual TimestampResolution TimestampPrecision
+        {
+            get => throw new NotSupportedException();
```
Comment (kayoub5) [inline SharpPcap/WinDivert/WinDivertDevice.cs:339]:
```suggestion
            get => TimestampResolution.Microsecond;
```

**Review by kayoub5 — COMMENTED** [review]
The current `PosixTimeval` still does not support nano seconds

File: Test/PcapDeviceTest.cs [hunk 6]
```diff
@@ -51,12 +60,82 @@ public void DeviceProperties([PcapDevices] DeviceFixture fixture)
...
+        /// <param name="fixture"></param>
+        [Test]
+        public void DeviceOpenWithTimestampType([PcapDevices] DeviceFixture fixture)
+        {
+            using var device = (PcapDevice)fixture.GetDevice();
+
+            // open the device to retrieve the supported timestamps
+            device.Open();
+            var timestampTypes = device.TimestampsSupported;
+            device.Close();
```
Comment (kayoub5) [inline Test/PcapDeviceTest.cs:127]: Having to open and close device to get list of supported timestamps sounds counter intuitive.

File: Test/PcapDeviceTest.cs [hunk 7]
```diff
@@ -51,12 +60,82 @@ public void DeviceProperties([PcapDevices] DeviceFixture fixture)
...
+
+            Assert.IsNotEmpty(timestampTypes);
+
+            foreach(var timestampType in timestampTypes)
+            {
+                var configuration = new DeviceConfiguration();
+                configuration.TimestampType = timestampType;
+                device.Open(configuration);
```
Comment (kayoub5) [inline Test/PcapDeviceTest.cs:135]: this won't run for all timestamps, since you don't close the device at the end of each iteration

File: Test/PcapDeviceTest.cs [hunk 8]
```diff
@@ -51,12 +60,82 @@ public void DeviceProperties([PcapDevices] DeviceFixture fixture)
...
+        public void DeviceOpenWithTimestampPrecision([PcapDevices] DeviceFixture fixture)
+        {
+            using var device = (PcapDevice)fixture.GetDevice();
+
+            foreach (TimestampResolution resolution in Enum.GetValues(typeof(TimestampResolution)))
+            {
+                var configuration = new DeviceConfiguration();
+                configuration.TimestampResolution = resolution;
+                device.Open(configuration);
+                DeviceTest(device, resolution);
```
Comment (kayoub5) [inline Test/PcapDeviceTest.cs:111]: this won't run for all timestamps, since you don't close the device at the end of each iteration

File: Test/PcapDeviceTest.cs [hunk 9]
```diff
@@ -34,12 +34,21 @@ namespace Test
...
+        /// </summary>
         [Test]
-        public void DeviceProperties([PcapDevices] DeviceFixture fixture)
+        public void TimestampConversions()
         {
-            using var device = (PcapDevice)fixture.GetDevice();
+            foreach(TimestampType timestampType in Enum.GetValues(typeof(TimestampType)))
+            {
+                Assert.IsNotEmpty(PcapDevice.TimestampName(timestampType));
+                Assert.IsNotEmpty(PcapDevice.TimestampDescription(timestampType));
```
Comment (kayoub5) [inline Test/PcapDeviceTest.cs:46]: I think we could make the timestamp API more object oriented. Proposal:
Create class `PcapClock`, that describes and encapsulates `TimestampType`:
```cs
public class PcapClock {
    public string Name { get; }
    public string Description { get; }
    public TimestampPrecision Precision {get; } // High, Low, Unknown
    public bool? Synchronized { get; }
}
```
this provide easier distinction between different timestamp types.

File: SharpPcap/ICaptureDevice.cs (hunk 1 above)
Comment (chmorgan) [reply SharpPcap/ICaptureDevice.cs:175]: it should be available for all pcap devices, so live, file reader and writer.

You noticed the precision vs. resolution usage too? It's all of the place in libpcap docs and code.

So the function uses precision, https://github.com/the-tcpdump-group/libpcap/blob/master/pcap/pcap.h#L428

but they do use resolution in some language. The variables are called precision though so let me rename appropriately. It's confusing inside of their documentation for sure. And language wise they are not the same thing. I think 'resolution' is the correct term though as this is the smallest unit vs. precision which is how close the values are to the actual value.

If we standardize on what libpcap uses that would mean using 'precision'. Properly I think we should use 'resolution' a s its the correct term.

File: SharpPcap/TimestampType.cs (hunk 4 above)
Comment (chmorgan) [reply SharpPcap/TimestampType.cs:11]: maybe assign the first and let the compiler assign the rest? They are in order at the moment.

File: Test/PcapDeviceTest.cs (hunk 6 above)
Comment (chmorgan) [reply Test/PcapDeviceTest.cs:127]: That's how they roll over there in libpcap land, see the function we use to retrieve the list, https://www.tcpdump.org/manpages/pcap_list_tstamp_types.3pcap.html

File: SharpPcap/TimestampType.cs (hunk 4 above)
Comment (kayoub5) [reply SharpPcap/TimestampType.cs:11]: better be explicit to prevent someone from adding one in the middle and shifting the value of the others.

File: Test/PcapDeviceTest.cs (hunk 6 above)
Comment (kayoub5) [reply Test/PcapDeviceTest.cs:127]: we can hide this complexity by creating a temporary handle internally.

File: Test/PcapDeviceTest.cs (hunk 9 above)
Comment (chmorgan) [reply Test/PcapDeviceTest.cs:46]: I think this makes sense. If we rename the existing precision to resolution what do we use for the types of resolution here?

File: Test/PcapDeviceTest.cs (hunk 6 above)
Comment (chmorgan) [reply Test/PcapDeviceTest.cs:127]: I don't think there is anything wrong with opening and closing devices or just creating the handle and using that. Yeah let me fix.

File: SharpPcap/LibPcap/PcapDevice.cs (hunk 3 above)
Comment (chmorgan) [reply SharpPcap/LibPcap/PcapDevice.cs:592]: we should throw I think. I know its not recommended to throw from properties but we aren't the normal class since we interact with hardware. I'm ok if we break that rule vs. making a method.

File: SharpPcap/ICaptureDevice.cs (hunk 1 above)
Comment (kayoub5) [reply SharpPcap/ICaptureDevice.cs:175]: @chmorgan there one detail I could clarify:
* resolution set by `PCAP_TSTAMP_PRECISION_NANO` affects how the timestamp `timeval` should be parsed (the microsecond field becomes a nanosecond field) 
* All other timestamp api affects how libpcap gets the timestamp but not how you should parse it.

Comment (kayoub5) [reply SharpPcap/ICaptureDevice.cs:175]: @chmorgan Now I think about it, since `pcap_set_tstamp_precision` changes how the timestamp value should be parsed.
It would be better to split this PR into two smaller ones:
* one for timestamp type `pcap_set_tstamp_type` and related functions
* one for timestamp resolution `pcap_set_tstamp_precision` and related functions, and updating `PosixTimeval`

File: SharpPcap/LibPcap/PcapDevice.cs [hunk 10]
```diff
@@ -586,6 +586,51 @@ public virtual void SendPacket(byte[] p, int size)
...
+            set => LibPcapSafeNativeMethods.pcap_set_tstamp_precision(PcapHandle, (int)value);
```
Comment (kayoub5) [inline SharpPcap/LibPcap/PcapDevice.cs:593]: @chmorgan Are you sure this setter actually works?
`pcap_set_tstamp_precision` can not be called on an activated pcap handle.

File: SharpPcap/ICaptureDevice.cs (hunk 1 above)
Comment (chmorgan) [reply SharpPcap/ICaptureDevice.cs:175]: Yep. Well its more about how the user would parse it. My thought, even though it isn't implemented yet, is to have a flag on PosixTimeval to indicate which resolution it has and expose a tuple of either (second, microsecond) or (second, nanosecond) that can be retrieved. We could have separate types for each, like PosixTimevalMS or PosixTimevalNS but I wasn't sure that would help the user as much.

File: SharpPcap/LibPcap/PcapDevice.cs (hunk 3 above)
Comment (kayoub5) [reply SharpPcap/LibPcap/PcapDevice.cs:592]: Let me rephrase: what happens when you do `LibPcapSafeNativeMethods.pcap_get_tstamp_precision(IntPtr.Zero)` ?

File: SharpPcap/TimestampType.cs (hunk 4 above)
Comment (chmorgan) [reply SharpPcap/TimestampType.cs:11]: We'd be the ones adding it or reviewing it though. Let me make sure I link back to the source code where that array is defined. I guess I'm less concerned that we'll thing we can just reorder that list. The whole project is based upon a proper mapping of various fields to specific orders etc.

File: SharpPcap/LibPcap/PcapDevice.cs (hunk 10 above)
Comment (chmorgan) [reply SharpPcap/LibPcap/PcapDevice.cs:593]: yeah that's probably broken, let me remove and comment on usage

File: SharpPcap/LibPcap/PcapDevice.cs [hunk 11]
```diff
@@ -586,6 +586,51 @@ public virtual void SendPacket(byte[] p, int size)
...
+        }
+
+        public List<TimestampType> TimestampsSupported
+        {
+            get
+            {
+                IntPtr typePtr = IntPtr.Zero;
+
+                // Note: typePtr must be freed with pcap_free_tstamp_types()
+                var typeCount = LibPcapSafeNativeMethods.pcap_list_tstamp_types(PcapHandle, ref typePtr);
```
Comment (kayoub5) [inline SharpPcap/LibPcap/PcapDevice.cs:616]: @chmorgan if `typeCount == 0`, return a list containing only the element `PCAP_TSTAMP_HOST`
See https://www.tcpdump.org/manpages/pcap_list_tstamp_types.3pcap.html

Comment (chmorgan) [reply SharpPcap/LibPcap/PcapDevice.cs:616]: good catch!

File: SharpPcap/LibPcap/PcapDevice.cs (hunk 3 above)
Comment (chmorgan) [reply SharpPcap/LibPcap/PcapDevice.cs:592]: I'd guess bad things occur, we should check and throw there.

File: Test/PcapDeviceTest.cs (hunk 6 above)
Comment (chmorgan) [reply Test/PcapDeviceTest.cs:127]: @kayoub5 since we can retrieve timestamps for any kind of device I'm not sure how this maps to things like capture reader and capture writer devices. At the PcapDevice level we don't know what kind of device it is. I did add a check to throw if not open though, which is necessary.

Comment (kayoub5) [reply Test/PcapDeviceTest.cs:129]: what about adding this api to live device class instead of the base class here?

**Comment by kayoub5** [comment]
@chmorgan may add commits to this pr?

File: Test/PcapDeviceTest.cs (hunk 6 above)
Comment (chmorgan) [reply Test/PcapDeviceTest.cs:129]: Then we'd leave out the offline devices, the file reader and writer, which can also have their own possible timestamps. I don't see a good path forward yet but I'm guessing there is one.

**Comment by chmorgan** [comment]
@kayoub5 if you want to add commits I guess so. I'm not opposed to seeing what you have in mind as long as you aren't opposed to me changing them when I get back to it tonight :-)

File: Test/PcapDeviceTest.cs (hunk 6 above)
Comment (kayoub5) [reply Test/PcapDeviceTest.cs:129]: file device can have a timestamp resolution, but not a timestamp type

Comment (chmorgan) [reply Test/PcapDeviceTest.cs:129]: ooh a very good spot, yeah they have their own functions for specifying the timestamp resolution.

Let me move that into the live device.


For capture reader and writer it seems like we could either add the option to the constructor or switch to having the user call Open() on the devices. The former is simpler for the user, the latter is more uniform with opening capture devices. I'm curious which one you think feels better, or do you have another approach?

**Comment by chmorgan** [comment]
@kayoub5 we do have a number of things to do. It would be more relaxing for me at least to not worry about racing or stepping over each other on the pr.

would it be ok if I just kept fixing and cleaning up the branch on my own?

We’ve got code coverage to improve, I have some arp tests coming as a heads up btw, and other improvements. Or did you run out of ideas for things? Heh

File: Test/PcapDeviceTest.cs (hunk 6 above)
Comment (kayoub5) [reply Test/PcapDeviceTest.cs:129]: If timestamp is the ONLY option available for files then simply adding it to the constructor would be simpler, if there are more options that could be configured, the `Open()` method would make more sense then.

A third option would be to make `Open()` static factory method as per the proposal in #201

Comment (chmorgan) [reply Test/PcapDeviceTest.cs:129]: @kayoub5 for the static factory approach I presume internally we'd preserve the classes and constructors and return the objects from the appropriate static factory method? It looks more tidy with Pcap.Openxxx() but how does it reduce effort/code/simplify syntax for the user over the current approach, now that we've paired down the Open() calls?

**Comment by chmorgan** [comment]
@kayoub5 I can split off the PcapException changes into their own PR if those look good but the rest doesn't look great.

Also had to put in place a handful of changes to keep the tests passing, for example by disabling them for WinPcap. I was thinking about adding a property to PcapDevice that would use the version to determine if the timestamp functions were supported, and using that to skip tests that used timestamp functions. Thoughts?

File: SharpPcap/LibPcap/CaptureFileReaderDevice.cs [hunk 12]
```diff
@@ -90,8 +90,26 @@ public override void Open(DeviceConfiguration configuration)
...
+            if (offline_with_tstamp_precision)
+            {
+                adapterHandle = LibPcapSafeNativeMethods.pcap_open_offline_with_tstamp_precision(m_pcapFile, (uint)resolution, errbuf);
+            } else
+            {
+                // notify the user that they asked for a non-standard resolution but their libpcap
+                // version lacks the necessary function
+                if(resolution != TimestampResolution.Microsecond)
+                {
+                    throw new NotSupportedException("pcap version is < 1.5.1 and TimeStampResolution != Microsecond, needs pcap_open_offline_with_tstamp_precision()");
```
Comment (kayoub5) [inline SharpPcap/LibPcap/CaptureFileReaderDevice.cs:108]: configuration.RaiseConfigurationFailed(
                            nameof(configuration.TimeStampResolution),
                            new NotSupportedException("pcap version is < 1.5.1")
                        );

File: SharpPcap/LibPcap/CaptureFileWriterDevice.cs [hunk 13]
```diff
@@ -117,7 +117,15 @@ public override void Open(DeviceConfiguration configuration)
             }
 
             // set the device handle
-            PcapHandle = LibPcapSafeNativeMethods.pcap_open_dead((int)configuration.LinkLayerType, configuration.Snaplen);
+            if (configuration.TimestampResolution.HasValue)
+            {
+                PcapHandle = LibPcapSafeNativeMethods.pcap_open_dead_with_tstamp_precision((int)configuration.LinkLayerType,
```
Comment (kayoub5) [inline SharpPcap/LibPcap/CaptureFileWriterDevice.cs:122]: No check for `Pcap.LibpcapVersion >= new Version(1, 5, 1)` ?

File: SharpPcap/LibPcap/LibPcapLiveDevice.cs [hunk 14]
```diff
@@ -349,5 +365,57 @@ public override ICaptureStatistics Statistics
...
+#region Timestamp
+        /// <summary>
+        /// Timestamps supported by this device
+        /// </summary>
+        ///
+        /// <remarks>
+        /// Note: Live devices can have supported timestamps but offline devices
+        /// (such as file readers etc) do not. See https://www.tcpdump.org/manpages/pcap-tstamp.7.html
+        /// </remarks>
+        public System.Collections.Generic.List<PcapClock> TimestampsSupported
```
Comment (kayoub5) [inline SharpPcap/LibPcap/LibPcapLiveDevice.cs:378]: I was thinking that this may go better in the `PcapInterface` class, instead of here. thoughts ?

File: SharpPcap/LibPcap/PcapDevice.cs [hunk 15]
```diff
@@ -586,6 +586,38 @@ public virtual void SendPacket(byte[] p, int size)
...
+            var ptr =  LibPcapSafeNativeMethods.pcap_tstamp_type_val_to_name((int)number);
+            return Marshal.PtrToStringAnsi(ptr);
+
+        }
+
+        public static string TimestampDescription(TimestampType number)
+        {
+            var ptr = LibPcapSafeNativeMethods.pcap_tstamp_type_val_to_description((int)number);
+            return Marshal.PtrToStringAnsi(ptr);
+        }
```
Comment (kayoub5) [inline SharpPcap/LibPcap/PcapDevice.cs:618]: I think it would be better to put `TimestampName(TimestampType number)` and `TimestampDescription(TimestampType number)` in `PcapClock` class instead (not public). thoughts?

File: SharpPcap/PcapClock.cs [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+        /// </summary>
+        public TimestampType TimestampType { get; }
+
+
+        public PcapClock(TimestampType timestampType, string name, string description)
+        {
+            Name = name;
+            Description = description;
+
+            Resolution = (timestampType == TimestampType.HostLowPrecision) ? TimestampSourceResolution.Low : TimestampSourceResolution.High;
```
Comment (kayoub5) [inline SharpPcap/PcapClock.cs:26]: if TimestampType == TimestampType.Host then Precision = Unknown

File: SharpPcap/TimestampSourceResolution.cs [hunk 17]
```diff
@@ -0,0 +1,10 @@
+﻿using System;
+namespace SharpPcap
+{
+    public enum TimestampSourceResolution
+    {
+        Low,
+        Medium,
```
Comment (kayoub5) [inline SharpPcap/TimestampSourceResolution.cs:7]:
```suggestion
        Unknown,
```

File: SharpPcap/PcapClock.cs [hunk 18]
```diff
@@ -0,0 +1,32 @@
...
+            Synchronized = ((timestampType == TimestampType.AdapterUnsynced) ||
```
Comment (kayoub5) [inline SharpPcap/PcapClock.cs:27]: if TimestampType == TimestampType.Host then Synchronized = null (unknown)

File: Test/PcapDeviceTest.cs [hunk 19]
```diff
@@ -52,9 +62,65 @@ public void DeviceProperties([PcapDevices] DeviceFixture fixture)
...
+                    else
+                    {
+                        Console.WriteLine("Pass: {0} expected {1} and was {2}", device.Name, targetResolution, resolution);
+                    }
+
+                    // Note: This doesn't not appear to work for any platform so it is
+                    // disabled for now
+                    //                Assert.AreEqual(targetResolution, resolution);
+                }
+            } catch (System.EntryPointNotFoundException)
```
Comment (kayoub5) [inline Test/PcapDeviceTest.cs:84]: `EntryPointNotFoundException` should not propagte outside of the library, it should be wrapped into NotSupportedException

File: Test/PcapDeviceTest.cs [hunk 20]
```diff
@@ -35,12 +35,22 @@ namespace Test
     [NonParallelizable]
     public class PcapDeviceTest
     {
+        /// <summary>
+        /// Test that we can convert each TimestampType
+        /// </summary>
+        [Category("Timestamp")]
```
Comment (kayoub5) [inline Test/PcapDeviceTest.cs:41]: If you want, I can create a custom attribute that automatically disable/enable unit tests based on Libpcap version

File: SharpPcap/PcapClock.cs (hunk 18 above)
Comment (chmorgan) [reply SharpPcap/PcapClock.cs:27]: @kayoub5 that makes sense. So its unknown if we are synchronized or not if its host.

File: SharpPcap/LibPcap/CaptureFileWriterDevice.cs (hunk 13 above)
Comment (chmorgan) [reply SharpPcap/LibPcap/CaptureFileWriterDevice.cs:122]: Yeah we should probably check that and report it, let me fix.

File: SharpPcap/LibPcap/PcapDevice.cs (hunk 15 above)
Comment (chmorgan) [reply SharpPcap/LibPcap/PcapDevice.cs:618]: @kayoub5 in that case we'd have LibPcapLiveDevice (or PcapInterface) calling static methods in PcapClock to get values that we'd then pass back to PcapClock? That feels a bit circular no?

File: SharpPcap/LibPcap/LibPcapLiveDevice.cs (hunk 14 above)
Comment (chmorgan) [reply SharpPcap/LibPcap/LibPcapLiveDevice.cs:378]: While the pcap pages do indicate that only capture devices support these calls I'm guessing, but haven't tested, that it won't have some exception. Let me see about moving it over. I can't say I have a good feel for whether its an improvement to do so yet.

File: SharpPcap/LibPcap/PcapDevice.cs (hunk 15 above)
Comment (kayoub5) [reply SharpPcap/LibPcap/PcapDevice.cs:618]: My idea was that PcapClock will only have one argument in the constructor, the TimestampType, and the logic for getting the value of all other properties would be encapsulated in PcapClock

File: SharpPcap/TimestampSourceResolution.cs (hunk 17 above)
Comment (kayoub5) [reply SharpPcap/TimestampSourceResolution.cs:7]: `Medium` timestamp precision does not exist

File: Test/PcapDeviceTest.cs (hunk 20 above)
Comment (kayoub5) [reply Test/PcapDeviceTest.cs:41]: See #233

File: Test/PcapDeviceTest.cs [hunk 21]
```diff
@@ -52,9 +64,65 @@ public void DeviceProperties([PcapDevices] DeviceFixture fixture)
...
+                        Console.WriteLine("Error: {0} expected {1} but was {2}", device.Name, targetResolution, resolution);
+                    }
+                    else
+                    {
+                        Console.WriteLine("Pass: {0} expected {1} and was {2}", device.Name, targetResolution, resolution);
+                    }
+
+                    // Note: This doesn't not appear to work for any platform so it is
+                    // disabled for now
+                    //                Assert.AreEqual(targetResolution, resolution);
```
Comment (kayoub5) [inline Test/PcapDeviceTest.cs:84]: Is this still not working?

File: SharpPcap/TimestampSourceResolution.cs (hunk 17 above)
Comment (chmorgan) [reply SharpPcap/TimestampSourceResolution.cs:7]: You are correct. I thought I saw it in a header file but I likely just made it up. hah Fixed

File: Test/PcapDeviceTest.cs (hunk 21 above)
Comment (chmorgan) [reply Test/PcapDeviceTest.cs:84]: let me test here again

Comment (chmorgan) [reply Test/PcapDeviceTest.cs]: works here, pushing for a CI retest

**Review by kayoub5 — APPROVED** [review]
